### PR TITLE
Move random-related defines from e_os.h to rand_unix.c

### LINKED ISSUE #10049: refactor random-related stuff out of e_os.h
I think only the crypto/rand files need the random stuff, so you can move 45 lines out of e_os.h to a separate file, or perhaps crypto/rand/rand_local.h ?

### PR DESCRIPTION
**Pull request by mspncp**
Fixes #10049

### CODE REVIEW

File: e_os.h [hunk 1]
```diff
@@ -17,56 +17,6 @@
 # include <openssl/crypto.h>
 # include "internal/nelem.h"
 
-/*
- * <openssl/e_os2.h> contains what we can justify to make visible to the
- * outside; this file e_os.h is not part of the exported interface.
- */
```
Comment (levitte) [inline e_os.h]: Do not remove this comment

Comment (mspncp) [reply e_os.h]: Yeah, your're right. I did not look close enough when moving the block.

**Comment by mspncp** [comment]
> I would even move that section of code to `rand_unix.c`. It's only relevant there.

Ok

File: crypto/rand/rand_unix.c [hunk 2]
```diff
@@ -81,7 +81,8 @@ static uint64_t get_timer_bits(void);
 #   define OSSL_POSIX_TIMER_OKAY
 #  endif
 # endif
-#endif /* defined(OPENSSL_SYS_UNIX) || defined(__DJGPP__) */
+#endif /* (defined(OPENSSL_SYS_UNIX) && !defined(OPENSSL_SYS_VXWORKS))
+          || defined(__DJGPP__) */
```
Comment (mspncp) [inline crypto/rand/rand_unix.c]: _FWIW @levitte: when checking #ifdef indentation with Emacs, its lax interpretation of whitespace (\<space> matches any sequence of whitespace characters) can become quite annoying. To turn it off, you can add the following lines to your .emacs file:_

```
(setq isearch-lax-whitespace nil)
(setq isearch-regexp-lax-whitespace nil)
```

Comment (mspncp) [reply crypto/rand/rand_unix.c]: I'll complete the pull request later, as soon as I figured out the correct location for the typedefs from e_os.h in the #ifdef spaghetti of rand_unix.c

**Comment by mspncp** [comment]
> I'll complete the pull request later, as soon as I figured out the correct location for the typedefs from e_os.h in the #ifdef spaghetti of rand_unix.c

Sorry, I got interrupted with other things, which prevented me from untangling the #ifdef spaghetti.  I'll push a fixup soon.

_Just in case anybody is wondering what I'm talking about: this is the reason why I hate rand_unix.c:_

```make
crypto/rand/rand_unix.c:10:#ifndef _GNU_SOURCE
crypto/rand/rand_unix.c:11:# define _GNU_SOURCE
crypto/rand/rand_unix.c:12:#endif
crypto/rand/rand_unix.c:13:#include "e_os.h"
crypto/rand/rand_unix.c:14:#include <stdio.h>
crypto/rand/rand_unix.c:15:#include "internal/cryptlib.h"
crypto/rand/rand_unix.c:16:#include <openssl/rand.h>
crypto/rand/rand_unix.c:17:#include <openssl/crypto.h>
crypto/rand/rand_unix.c:18:#include "rand_local.h"
crypto/rand/rand_unix.c:19:#include "crypto/rand.h"
crypto/rand/rand_unix.c:20:#include <stdio.h>
crypto/rand/rand_unix.c:21:#include "internal/dso.h"
crypto/rand/rand_unix.c:22:#ifdef __linux
crypto/rand/rand_unix.c:23:# include <sys/syscall.h>
crypto/rand/rand_unix.c:24:# ifdef DEVRANDOM_WAIT
crypto/rand/rand_unix.c:25:#  include <sys/shm.h>
crypto/rand/rand_unix.c:26:#  include <sys/utsname.h>
crypto/rand/rand_unix.c:27:# endif
crypto/rand/rand_unix.c:28:#endif
crypto/rand/rand_unix.c:29:#if defined(__FreeBSD__) && !defined(OPENSSL_SYS_UEFI)
crypto/rand/rand_unix.c:30:# include <sys/types.h>
crypto/rand/rand_unix.c:31:# include <sys/sysctl.h>
crypto/rand/rand_unix.c:32:# include <sys/param.h>
crypto/rand/rand_unix.c:33:#endif
crypto/rand/rand_unix.c:34:#if defined(__OpenBSD__) || defined(__NetBSD__)
crypto/rand/rand_unix.c:35:# include <sys/param.h>
crypto/rand/rand_unix.c:36:#endif
crypto/rand/rand_unix.c:38:#if (defined(OPENSSL_SYS_UNIX) && !defined(OPENSSL_SYS_VXWORKS)) \
crypto/rand/rand_unix.c:40:# include <sys/types.h>
crypto/rand/rand_unix.c:41:# include <sys/stat.h>
crypto/rand/rand_unix.c:42:# include <fcntl.h>
crypto/rand/rand_unix.c:43:# include <unistd.h>
crypto/rand/rand_unix.c:44:# include <sys/time.h>
crypto/rand/rand_unix.c:50:# define TWO32TO64(a, b) ((((uint64_t)(a)) << 32) + (b))
crypto/rand/rand_unix.c:72:# undef OSSL_POSIX_TIMER_OKAY
crypto/rand/rand_unix.c:73:# if defined(_POSIX_TIMERS) && _POSIX_TIMERS > 0
crypto/rand/rand_unix.c:74:#  if defined(__GLIBC__)
crypto/rand/rand_unix.c:75:#   if defined(__GLIBC_PREREQ)
crypto/rand/rand_unix.c:76:#    if __GLIBC_PREREQ(2, 17)
crypto/rand/rand_unix.c:77:#     define OSSL_POSIX_TIMER_OKAY
crypto/rand/rand_unix.c:78:#    endif
crypto/rand/rand_unix.c:79:#   endif
crypto/rand/rand_unix.c:80:#  else
crypto/rand/rand_unix.c:81:#   define OSSL_POSIX_TIMER_OKAY
crypto/rand/rand_unix.c:82:#  endif
crypto/rand/rand_unix.c:83:# endif
crypto/rand/rand_unix.c:84:#endif /* (defined(OPENSSL_SYS_UNIX) && !defined(OPENSSL_SYS_VXWORKS))
crypto/rand/rand_unix.c:87:#if defined(OPENSSL_RAND_SEED_NONE)
crypto/rand/rand_unix.c:89:# undef OPENSSL_RAND_SEED_OS
crypto/rand/rand_unix.c:90:# undef OPENSSL_RAND_SEED_GETRANDOM
crypto/rand/rand_unix.c:91:# undef OPENSSL_RAND_SEED_LIBRANDOM
crypto/rand/rand_unix.c:92:# undef OPENSSL_RAND_SEED_DEVRANDOM
crypto/rand/rand_unix.c:93:# undef OPENSSL_RAND_SEED_RDTSC
crypto/rand/rand_unix.c:94:# undef OPENSSL_RAND_SEED_RDCPU
crypto/rand/rand_unix.c:95:# undef OPENSSL_RAND_SEED_EGD
crypto/rand/rand_unix.c:96:#endif
crypto/rand/rand_unix.c:98:#if defined(OPENSSL_SYS_UEFI) && !defined(OPENSSL_RAND_SEED_NONE)
crypto/rand/rand_unix.c:99:# error "UEFI only supports seeding NONE"
crypto/rand/rand_unix.c:100:#endif
crypto/rand/rand_unix.c:102:#if !(defined(OPENSSL_SYS_WINDOWS) || defined(OPENSSL_SYS_WIN32) \
crypto/rand/rand_unix.c:106:# if defined(OPENSSL_SYS_VOS)
crypto/rand/rand_unix.c:108:#  ifndef OPENSSL_RAND_SEED_OS
crypto/rand/rand_unix.c:109:#   error "Unsupported seeding method configured; must be os"
crypto/rand/rand_unix.c:110:#  endif
crypto/rand/rand_unix.c:112:#  if defined(OPENSSL_SYS_VOS_HPPA) && defined(OPENSSL_SYS_VOS_IA32)
crypto/rand/rand_unix.c:113:#   error "Unsupported HP-PA and IA32 at the same time."
crypto/rand/rand_unix.c:114:#  endif
crypto/rand/rand_unix.c:115:#  if !defined(OPENSSL_SYS_VOS_HPPA) && !defined(OPENSSL_SYS_VOS_IA32)
crypto/rand/rand_unix.c:116:#   error "Must have one of HP-PA or IA32"
crypto/rand/rand_unix.c:117:#  endif
crypto/rand/rand_unix.c:140:#  ifdef OPENSSL_SYS_VOS_HPPA
crypto/rand/rand_unix.c:143:#  else
crypto/rand/rand_unix.c:146:#  endif
crypto/rand/rand_unix.c:158:#  ifdef OPENSSL_SYS_VOS_HPPA
crypto/rand/rand_unix.c:162:#  else
crypto/rand/rand_unix.c:166:#  endif
crypto/rand/rand_unix.c:184:# else
crypto/rand/rand_unix.c:186:#  if defined(OPENSSL_RAND_SEED_EGD) && \
crypto/rand/rand_unix.c:188:#   error "Seeding uses EGD but EGD is turned off or no device given"
crypto/rand/rand_unix.c:189:#  endif
crypto/rand/rand_unix.c:191:#  if defined(OPENSSL_RAND_SEED_DEVRANDOM) && !defined(DEVRANDOM)
crypto/rand/rand_unix.c:192:#   error "Seeding uses urandom but DEVRANDOM is not configured"
crypto/rand/rand_unix.c:193:#  endif
crypto/rand/rand_unix.c:195:#  if defined(OPENSSL_RAND_SEED_OS)
crypto/rand/rand_unix.c:196:#   if !defined(DEVRANDOM)
crypto/rand/rand_unix.c:197:#    error "OS seeding requires DEVRANDOM to be configured"
crypto/rand/rand_unix.c:198:#   endif
crypto/rand/rand_unix.c:199:#   define OPENSSL_RAND_SEED_GETRANDOM
crypto/rand/rand_unix.c:200:#   define OPENSSL_RAND_SEED_DEVRANDOM
crypto/rand/rand_unix.c:201:#  endif
crypto/rand/rand_unix.c:203:#  if defined(OPENSSL_RAND_SEED_LIBRANDOM)
crypto/rand/rand_unix.c:204:#   error "librandom not (yet) supported"
crypto/rand/rand_unix.c:205:#  endif
crypto/rand/rand_unix.c:207:#  if (defined(__FreeBSD__) || defined(__NetBSD__)) && defined(KERN_ARND)
crypto/rand/rand_unix.c:241:#if   defined(__NetBSD__) && __NetBSD_Version__ < 400000000
crypto/rand/rand_unix.c:244:#endif
crypto/rand/rand_unix.c:260:#  endif
crypto/rand/rand_unix.c:262:#  if defined(OPENSSL_RAND_SEED_GETRANDOM)
crypto/rand/rand_unix.c:264:#   if defined(__linux) && !defined(__NR_getrandom)
crypto/rand/rand_unix.c:265:#    if defined(__arm__)
crypto/rand/rand_unix.c:266:#     define __NR_getrandom    (__NR_SYSCALL_BASE+384)
crypto/rand/rand_unix.c:267:#    elif defined(__i386__)
crypto/rand/rand_unix.c:268:#     define __NR_getrandom    355
crypto/rand/rand_unix.c:269:#    elif defined(__x86_64__)
crypto/rand/rand_unix.c:270:#     if defined(__ILP32__)
crypto/rand/rand_unix.c:271:#      define __NR_getrandom   (__X32_SYSCALL_BIT + 318)
crypto/rand/rand_unix.c:272:#     else
crypto/rand/rand_unix.c:273:#      define __NR_getrandom   318
crypto/rand/rand_unix.c:274:#     endif
crypto/rand/rand_unix.c:275:#    elif defined(__xtensa__)
crypto/rand/rand_unix.c:276:#     define __NR_getrandom    338
crypto/rand/rand_unix.c:277:#    elif defined(__s390__) || defined(__s390x__)
crypto/rand/rand_unix.c:278:#     define __NR_getrandom    349
crypto/rand/rand_unix.c:279:#    elif defined(__bfin__)
crypto/rand/rand_unix.c:280:#     define __NR_getrandom    389
crypto/rand/rand_unix.c:281:#    elif defined(__powerpc__)
crypto/rand/rand_unix.c:282:#     define __NR_getrandom    359
crypto/rand/rand_unix.c:283:#    elif defined(__mips__) || defined(__mips64)
crypto/rand/rand_unix.c:284:#     if _MIPS_SIM == _MIPS_SIM_ABI32
crypto/rand/rand_unix.c:285:#      define __NR_getrandom   (__NR_Linux + 353)
crypto/rand/rand_unix.c:286:#     elif _MIPS_SIM == _MIPS_SIM_ABI64
crypto/rand/rand_unix.c:287:#      define __NR_getrandom   (__NR_Linux + 313)
crypto/rand/rand_unix.c:288:#     elif _MIPS_SIM == _MIPS_SIM_NABI32
crypto/rand/rand_unix.c:289:#      define __NR_getrandom   (__NR_Linux + 317)
crypto/rand/rand_unix.c:290:#     endif
crypto/rand/rand_unix.c:291:#    elif defined(__hppa__)
crypto/rand/rand_unix.c:292:#     define __NR_getrandom    (__NR_Linux + 339)
crypto/rand/rand_unix.c:293:#    elif defined(__sparc__)
crypto/rand/rand_unix.c:294:#     define __NR_getrandom    347
crypto/rand/rand_unix.c:295:#    elif defined(__ia64__)
crypto/rand/rand_unix.c:296:#     define __NR_getrandom    1339
crypto/rand/rand_unix.c:297:#    elif defined(__alpha__)
crypto/rand/rand_unix.c:298:#     define __NR_getrandom    511
crypto/rand/rand_unix.c:299:#    elif defined(__sh__)
crypto/rand/rand_unix.c:300:#     if defined(__SH5__)
crypto/rand/rand_unix.c:301:#      define __NR_getrandom   373
crypto/rand/rand_unix.c:302:#     else
crypto/rand/rand_unix.c:303:#      define __NR_getrandom   384
crypto/rand/rand_unix.c:304:#     endif
crypto/rand/rand_unix.c:305:#    elif defined(__avr32__)
crypto/rand/rand_unix.c:306:#     define __NR_getrandom    317
crypto/rand/rand_unix.c:307:#    elif defined(__microblaze__)
crypto/rand/rand_unix.c:308:#     define __NR_getrandom    385
crypto/rand/rand_unix.c:309:#    elif defined(__m68k__)
crypto/rand/rand_unix.c:310:#     define __NR_getrandom    352
crypto/rand/rand_unix.c:311:#    elif defined(__cris__)
crypto/rand/rand_unix.c:312:#     define __NR_getrandom    356
crypto/rand/rand_unix.c:313:#    elif defined(__aarch64__)
crypto/rand/rand_unix.c:314:#     define __NR_getrandom    278
crypto/rand/rand_unix.c:315:#    else /* generic */
crypto/rand/rand_unix.c:316:#     define __NR_getrandom    278
crypto/rand/rand_unix.c:317:#    endif
crypto/rand/rand_unix.c:318:#   endif
crypto/rand/rand_unix.c:346:#  if defined(__GNUC__) && __GNUC__>=2 && defined(__ELF__) && !defined(__hpux)
crypto/rand/rand_unix.c:351:#  elif !defined(FIPS_MODE)
crypto/rand/rand_unix.c:366:#  endif
crypto/rand/rand_unix.c:369:#  if defined(__linux) && defined(__NR_getrandom)
crypto/rand/rand_unix.c:371:#  elif (defined(__FreeBSD__) || defined(__NetBSD__)) && defined(KERN_ARND)
crypto/rand/rand_unix.c:373:#  else
crypto/rand/rand_unix.c:376:#  endif
crypto/rand/rand_unix.c:378:#  endif    /* defined(OPENSSL_RAND_SEED_GETRANDOM) */
crypto/rand/rand_unix.c:380:#  if defined(OPENSSL_RAND_SEED_DEVRANDOM)
crypto/rand/rand_unix.c:391:#   if defined(__linux) && defined(DEVRANDOM_WAIT)
crypto/rand/rand_unix.c:394:#    if !defined(FIPS_MODE)
crypto/rand/rand_unix.c:399:#    endif
crypto/rand/rand_unix.c:465:#    ifndef FIPS_MODE
crypto/rand/rand_unix.c:469:#    endif
crypto/rand/rand_unix.c:474:#   else /* defined __linux */
crypto/rand/rand_unix.c:479:#   endif
crypto/rand/rand_unix.c:568:#  else     /* !defined(OPENSSL_RAND_SEED_DEVRANDOM) */
crypto/rand/rand_unix.c:583:#  endif    /* defined(OPENSSL_RAND_SEED_DEVRANDOM) */
crypto/rand/rand_unix.c:604:#  if defined(OPENSSL_RAND_SEED_NONE)
crypto/rand/rand_unix.c:606:#  else
crypto/rand/rand_unix.c:609:#   if defined(OPENSSL_RAND_SEED_GETRANDOM)
crypto/rand/rand_unix.c:633:#   endif
crypto/rand/rand_unix.c:635:#   if defined(OPENSSL_RAND_SEED_LIBRANDOM)
crypto/rand/rand_unix.c:639:#   endif
crypto/rand/rand_unix.c:641:#   if defined(OPENSSL_RAND_SEED_DEVRANDOM)
crypto/rand/rand_unix.c:679:#   endif
crypto/rand/rand_unix.c:681:#   if defined(OPENSSL_RAND_SEED_RDTSC)
crypto/rand/rand_unix.c:685:#   endif
crypto/rand/rand_unix.c:687:#   if defined(OPENSSL_RAND_SEED_RDCPU)
crypto/rand/rand_unix.c:691:#   endif
crypto/rand/rand_unix.c:693:#   if defined(OPENSSL_RAND_SEED_EGD)
crypto/rand/rand_unix.c:718:#   endif
crypto/rand/rand_unix.c:721:#  endif
crypto/rand/rand_unix.c:723:# endif
crypto/rand/rand_unix.c:724:#endif
crypto/rand/rand_unix.c:726:#if (defined(OPENSSL_SYS_UNIX) && !defined(OPENSSL_SYS_VXWORKS)) \
crypto/rand/rand_unix.c:785:# if defined(OSSL_POSIX_TIMER_OKAY)
crypto/rand/rand_unix.c:792:# endif
crypto/rand/rand_unix.c:793:# if defined(__unix__) \
crypto/rand/rand_unix.c:801:# endif
crypto/rand/rand_unix.c:819:# if defined(__sun) || defined(__hpux)
crypto/rand/rand_unix.c:821:# elif defined(_AIX)
crypto/rand/rand_unix.c:828:# elif defined(OSSL_POSIX_TIMER_OKAY)
crypto/rand/rand_unix.c:832:#  ifdef CLOCK_BOOTTIME
crypto/rand/rand_unix.c:833:#   define CLOCK_TYPE CLOCK_BOOTTIME
crypto/rand/rand_unix.c:834:#  elif defined(_POSIX_MONOTONIC_CLOCK)
crypto/rand/rand_unix.c:835:#   define CLOCK_TYPE CLOCK_MONOTONIC
crypto/rand/rand_unix.c:836:#  else
crypto/rand/rand_unix.c:837:#   define CLOCK_TYPE CLOCK_REALTIME
crypto/rand/rand_unix.c:838:#  endif
crypto/rand/rand_unix.c:843:# endif
crypto/rand/rand_unix.c:844:# if defined(__unix__) \
crypto/rand/rand_unix.c:852:# endif
crypto/rand/rand_unix.c:855:#endif /* (defined(OPENSSL_SYS_UNIX) && !defined(OPENSSL_SYS_VXWORKS))
```

**Comment by mspncp** [comment]
Just rebased  to fix the conflict for now.

**Comment by t8m** [comment]
> Sorry, I got interrupted with other things, which prevented me from untangling the #ifdef spaghetti. I'll push a fixup soon.
> 
> _Just in case anybody is wondering what I'm talking about: this is the reason why I hate rand_unix.c:_

It would be very beneficial to untangle this. However it will be quite hard and it will require careful review to not regress anything.

**Comment by mspncp** [comment]
> > Sorry, I got interrupted with other things, which prevented me from untangling the #ifdef spaghetti. I'll push a fixup soon.
> > _Just in case anybody is wondering what I'm talking about: this is the reason why I hate rand_unix.c:_
> 
> It would be very beneficial to untangle this. However it will be quite hard and it will require careful review to not regress anything.

Oh, I just noticed that my formulation was ambiguous: I wasn't talking about the big untangling (this surely needs to be done some time, but not in this pr), I was only talking about finding the right location inside the spaghetti, where the code from e_os.h fits best. :grin:

**Comment by richsalz** [comment]
You don't have to find the *best* place to put the code, just a *working* place.  For example, putting it right after the include's and before anything else should be fine.  (Why is an exercise for the reader)

**Comment by mspncp** [comment]
Thanks for your hints @richsalz. TBH, it was just an excuse to hide the fact that I was too much distracted with other things like organizing OMC meetings and reviewing other PRs, which prevented me from making progress here...

**Comment by mspncp** [comment]
> You don't have to find the _best_ place to put the code, just a _working_ place. For example, putting it right after the include's and before anything else should be fine. (Why is an exercise for the reader)

The situation was not that simple, I couldn't just place it "after the includes". I had to place in the middle of the includes, high enough to ensure that the shared memory headers get included: 

https://github.com/openssl/openssl/blob/604ee9fa4981f9166852735826249081ad26c359/crypto/rand/rand_unix.c#L74-L77

It's not ideal, but it works. _(And again the entanglement of rand_unix.c increases with every new piece which is added to it. Time to clean it up...)_

**Comment by richsalz** [comment]
Sigh.  I did so much work to clean up the ifdef nonsense back in the day.

**Comment by mspncp** [comment]
> Sigh. I did so much work to clean up the ifdef nonsense back in the day.

What an irony that the second law of thermodynamics also applies to rand_unix.c: The more entropy (=mess) the file contains, the more it increases with every new piece you add.

**Comment by mspncp** [comment]
Maybe we should feed the excessive #ifdef entropy into the DRBG? 🤣

File: e_os.h (hunk 1 above)
Comment (mspncp) [reply e_os.h]: (resolved: comment was restored some time ago)

**Comment by mspncp** [comment]
Merged, thanks!

**master**
01036e2afb Move random-related defines from e_os.h to rand_unix.c
2a7e6ed86b rand_unix.c: correct include guard comments

**1.1.1**
7b18d1a53f Move random-related defines from e_os.h to rand_unix.c
10cb54d75b rand_unix.c: correct include guard comments

**Comment by paulidale** [comment]
This is late but my preference would be reorganising the rand_XXX.c files.  For XXX in { unix, windows, vms }.  Instead of defining the randomness sources by _operating system_, could we instead define them by _source_.  I.e. rand_device.c, rand_rdrand.c, rand_rdtsc.c, ...  The #ifdef maze could then be mostly avoided.

**Comment by mspncp** [comment]
I had similar thoughts, too. But never the time to get anything started.

**Pull request closed without merging**